File: packages/cds-android/src/main/java/com/coinbase/cds/components/slidebutton/SlideButtonStyle.kt
package com.coinbase.cds.components.slidebutton

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import com.coinbase.cds.theme.CdsTheme

/** Resolved handle/content colors for a [SlideButtonVariant]. The track is always `bgSecondary`
 * regardless of variant -- only the handle recolors. */
@Immutable
internal data class SlideButtonColors(val container: Color, val content: Color)

@Composable
internal fun slideButtonColors(variant: SlideButtonVariant): SlideButtonColors {
    val colors = CdsTheme.colors
    return when (variant) {
        SlideButtonVariant.Primary -> SlideButtonColors(colors.bgPrimary, colors.fgInverse)
        SlideButtonVariant.Positive -> SlideButtonColors(colors.bgPositive, colors.fgInverse)
        SlideButtonVariant.Negative -> SlideButtonColors(colors.bgNegative, colors.fgInverse)
    }
}

/**
 * Resolved size-derived metrics for a [SlideButtonSize]. Deliberately has no `height` field: the
 * drag math needs the handle's collapsed width up front, before any layout pass, but rather than
 * take that as a number [SlideButton] measures its own track at layout time and derives it. Height
 * then falls out of [paddingY] plus line height, the same way
 * [com.coinbase.cds.components.button.Button] is never given a height either.
 *
 * There's no `handleInset` field either, for the same reason: a fixed token only happens to center
 * the icon inside the collapsed (idle) handle if it exactly matches half the leftover space, which
 * depends on the *measured* collapsed diameter -- so [SlideButton] computes that inset directly
 * from the same track measurement instead of guessing at a static value.
 */
@Immutable
internal data class SlideButtonMetrics(
    val paddingY: Dp,
    val labelPaddingX: Dp,
    val radius: Dp,
    val iconSize: Dp,
    val font: TextStyle,
)

@Composable
internal fun slideButtonMetrics(size: SlideButtonSize): SlideButtonMetrics {
    val space = CdsTheme.space
    val radius = CdsTheme.borderRadius
    val iconSize = CdsTheme.iconSize
    val font = CdsTheme.typography.headline
    return when (size) {
        SlideButtonSize.S -> SlideButtonMetrics(space.x1, space.x2, radius.radius700, iconSize.s, font)
        SlideButtonSize.M -> SlideButtonMetrics(space.x1_5, space.x2, radius.radius900, iconSize.m, font)
        SlideButtonSize.L -> SlideButtonMetrics(space.x2, space.x2, radius.radius900, iconSize.m, font)
    }
}

File: packages/cds-ios/Sources/Components/SlideButtonStyle.swift
import SwiftUI

/// Visual/semantic variant — the subset of ``ButtonVariant`` that makes sense for a
/// confirm-by-sliding action.
enum SlideButtonVariant { case primary, positive, negative }

/// Size tier (`s`/`m`/`l`).
enum SlideButtonSize { case s, m, l }

/// Resolved handle/content colors for a ``SlideButtonVariant``. The track is always
/// `bgSecondary`; only the handle recolors.
struct SlideButtonColors {
    let container: Color
    let content: Color
}

func slideButtonColors(_ variant: SlideButtonVariant, theme: CDSTheme) -> SlideButtonColors {
    // The handle mirrors the equivalent solid button variant, so derive from the button mapping
    // rather than re-listing the same tokens.
    let buttonVariant: ButtonVariant
    switch variant {
    case .primary: buttonVariant = .primary
    case .positive: buttonVariant = .positive
    case .negative: buttonVariant = .negative
    }
    let colors = buttonColors(buttonVariant, transparent: false, theme: theme)
    return SlideButtonColors(container: colors.container, content: colors.content)
}

/// Resolved size-derived metrics for a ``SlideButtonSize``. Height is not a field: it emerges
/// from `paddingY` plus the font's line height, and the collapsed handle is a square of that height.
struct SlideButtonMetrics {
    let paddingY: CGFloat
    let labelPaddingX: CGFloat
    let radius: CGFloat
    let iconSize: CGFloat
    let font: CDSTextStyle
}

func slideButtonMetrics(_ size: SlideButtonSize, theme: CDSTheme) -> SlideButtonMetrics {
    let space = theme.spacing
    let radius = theme.radius
    let icon = theme.iconSize
    switch size {
    case .s: return SlideButtonMetrics(paddingY: space.x1, labelPaddingX: space.x2, radius: radius.r700, iconSize: icon.s, font: .headline)
    case .m: return SlideButtonMetrics(paddingY: space.x1_5, labelPaddingX: space.x2, radius: radius.r900, iconSize: icon.m, font: .headline)
    case .l: return SlideButtonMetrics(paddingY: space.x2, labelPaddingX: space.x2, radius: radius.r900, iconSize: icon.m, font: .headline)
    }
}
